### ISSUE TITLE
B2: Panic for concurrent upload token use after network disconnect

### ISSUE THREAD

**Issue by meeee** [restic/restic#1207]
## Output of `restic version`

```
restic 0.7.1
compiled with go1.8.3 on linux/amd64
```

## How did you run restic exactly?

I'm using a wrapper script that sets environment variables and uses trickle to limit bandwidth:

```
export B2_ACCOUNT_ID=<restic_b2_account_id>
export B2_ACCOUNT_KEY=<restic_b2_account_key>
export RESTIC_REPOSITORY=b2:<bucket>:
export RESTIC_PASSWORD=<restic_password>

# Limit upload to 900 KB/s and download to 4500 KB/s using trickle
exec trickle -s -u 900 -d 4500 restic "$@"
```

Command line:

```
restic-wrapper backup <path>
```

## What backend/server/service did you use?

Backblaze B2

## Expected behavior

restic can recover from failed uploads with the proper parameters for B2.

## Actual behavior

After a few hours of uploading:

```
panic: Close: b2_upload_file: 400: more than one upload using auth token <token removed>

goroutine 96011 [running]:
restic/archiver.(*Archiver).saveChunk(0xc420b5b180, 0xc213e0, 0xc420076338, 0x3e24c8, 0x14bf48, 0x2beec6dc00000, 0xc42d740000, 0x14bf48, 0x18a000, 0xc42029e000, ...)
        src/restic/archiver/archiver.go:165 +0x395
created by restic/archiver.(*Archiver).SaveFile
        src/restic/archiver/archiver.go:239 +0x3a4
```

## Steps to reproduce the behavior

Run the command described above with a backup set of about 200 GB.

## Do you have any idea what may have caused this?

I suspect the nightly DSL reconnect caused this. The restic logs don't have timestamps, but the start time of a subsequent backup of another data set and the logs of my router support this suspicion.

My first guess would be this:

* Network connection breaks
* B2 doesn't realize an upload's connection is broken, assumes auth token is still in use
* restic notices the upload connection is broken (before B2 does) and starts a new upload with the same auth token
* B2 rejects the auth token for concurrent use

**Update:** Same error happened again this morning at the same time, so this seems to be reproducible.

## Do you have an idea how to solve the issue?

Instead of panicking, fetch a new auth token from B2 and continue uploading with that.

I haven't looked into the restic code, so this might actually be an issue with [blazer](https://github.com/kurin/blazer). Let me know if that's the case, then I'll report the issue there.

**Comment by fd0** [comment]
Hey, thanks for taking the time to report this. I think it's a bug in the `blazer` library (as you suspected). @kurin can you please take a look?

**Comment by kurin** [comment]
Ack, I'll look into it.

**Comment by kurin** [comment]
B2 upload auth tokens *ought* to be viable for 24 hours (https://www.backblaze.com/b2/docs/b2_get_upload_url.html, see the response).  I don't believe there are any restrictions on the use of this token.  In particular, from that page, "You can upload as many files to this URL as you need. To achieve faster upload speeds, request multiple uploadUrls and upload your files to these different endpoints in parallel."

B2 account auth tokens are similarly good for 24 hours (https://www.backblaze.com/b2/docs/b2_authorize_account.html) and I don't see how use of this token would result in the error returned.

Finally, this is an HTTP 400 error, which is not a good error to blindly retry on.

Which is all to say that I believe the *actual* bug is in B2, whether in their docs or implementation or both.  But that doesn't help us.

(Also, it appears this isn't unique to blazer: ncw/rclone#373)

The correct place to try to detect this is blazer, which already has code for detecting bad auth tokens and grabbing new ones (although there it can look for the more helpful--and documented--401 status code).  I'll see if I can get something out soon.  It would be helpful if you could get a (PII scrubbed!) log of the HTTP error response; maybe there's something better than a substring search of the error message I can do.

**Comment by kurin** [comment]
I guess a consistent interpretation with both the error and the documentation is that you can reuse the auth token as many times as you want in _serial_, but parallel use is verboten.  Still.  A 401 would be nice.

**Comment by kurin** [comment]
I have a theory about how to reliably reproduce it, and I'm trying to write a test for that, but it's going to involve writing an entire RoundTripper/Reader combo and yadda yadda yadda try https://github.com/kurin/blazer/tree/token for now and see if that helps.

**Comment by meeee** [comment]
@fd0 @kurin Thanks for looking into this so fast!

I restarted the backup with a debug build of v0.7.1 with the token branch of blazer and will report back.

**Comment by meeee** [comment]
@kurin The fix in the token branch works, thanks!

I've uploaded [log messages related to transport error messages, grouped by auth code](https://gist.github.com/meeee/c0b7684878d900867b3011d732a4a5fc).

The logs support my theory on why the error appears: For all four requests that fail with a 400, there are corresponding transport error messages (like 'connection reset') and a request a few seconds earlier that uses the same auth code. Both requests try to upload the same file. It's unclear why B2 doesn't notice that the connection is broken, though.

Unfortunately, the logs only contain HTTP headers. Let me know if a full response body of the error would still help you - I could try to manually force a reconnect and log the body.

**Comment by fd0** [comment]
That's great news!

**Comment by kurin** [comment]
I have a reliable test and I've released the fix in v0.1.1.

**Comment by fd0** [comment]
Wohoo, cool! Thank you! I've update the dependencies, see PR #1236.

**Comment by meeee** [comment]
@kurin @fd0 That was quick, thanks!

**Comment by fd0** [comment]
This is also included in the (just released) restic 0.7.2
